### ISSUE TITLE
Reading HttpServletRequest InputStream from a Filter then accessing getParameterNames() results in java.io.IOException: Missing content for multipart request

### ISSUE THREAD

**Issue by joakime** [eclipse/jetty.project#1175]
If you have a Filter that reads the HttpServletRequest.getInputStream(), then attempts to access the HttpServletRequest.getParameterNames() an IOException occurs.

(Stacktrace correlates to Jetty 9.4.0.v20161208 release)
```
java.lang.RuntimeException: java.io.IOException: Missing content for multipart request
	at org.eclipse.jetty.server.Request.extractMultipartParameters(Request.java:522)
	at org.eclipse.jetty.server.Request.extractContentParameters(Request.java:439)
	at org.eclipse.jetty.server.Request.getParameters(Request.java:363)
	at org.eclipse.jetty.server.Request.getParameterNames(Request.java:1015)
```

**Comment by joakime** [comment]
This was discovered due to the strange nature of tobago-2.0.x from issue #182 and #240.

The stacktrace as seen in tobago-2.0.x is as follows ...

```
java.io.IOException: Missing content for multipart request
	at org.eclipse.jetty.util.MultiPartInputStreamParser.parse(MultiPartInputStreamParser.java:554)
	at org.eclipse.jetty.util.MultiPartInputStreamParser.getParts(MultiPartInputStreamParser.java:441)
	at org.eclipse.jetty.server.Request.getParts(Request.java:2310)
	at org.eclipse.jetty.server.Request.extractMultipartParameters(Request.java:517)
	at org.eclipse.jetty.server.Request.extractContentParameters(Request.java:439)
	at org.eclipse.jetty.server.Request.getParameters(Request.java:363)
	at org.eclipse.jetty.server.Request.getParameterNames(Request.java:1015)
	at org.apache.myfaces.tobago.internal.webapp.TobagoMultipartFormdataRequest.init(TobagoMultipartFormdataRequest.java:128)
	at org.apache.myfaces.tobago.internal.webapp.TobagoMultipartFormdataRequest.<init>(TobagoMultipartFormdataRequest.java:63)
	at org.apache.myfaces.tobago.webapp.TobagoMultipartFormdataFilter.doFilter(TobagoMultipartFormdataFilter.java:113)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1613)
```

The process as seen in tobago-2.0.x ...

 1. `TobagoMultipartFormdataFilter.doFilter()` called
 2. `TobagoMultipartFormdataRequest` initialized with HttpServletRequest
 3. `commons-fileupload` is used to `ServletFileUpload.parseRequest(request)` to read the `HttpServletRequest.getInputStream()` completely
 4. `HttpServletRequest.getParameterNames()` is used next, which causes the IOException above

I feel that this is not a problem with the `MultiPartInputStreamParser`, but rather that the `Request.extractContentParameters()` not being aware that the InputStream has been completely read already.

Perhaps the `extractContentParameters()` should inspect the `_inputState==NONE` to verify proper behavior, and throw an `IllegalStateException` if an attempt to read/extract Parameters occurs after the input has been accessed be the application?

Either way, there's a bug in tobago-2.0.x's implementation too.

Either they get a successful commons-fileupload parse of the inputstream and an invalid Parameter map, or if they need a valid Parameter map, then they need to not use commons-fileupload parse, using `@MultiPartConfig` and `HttpServletRequest.getParts()` instead (which will handle parameter map parse correctly)

**Comment by joakime** [comment]
Added `@Ignored` [RequestTest#testMultiPartFormDataReadInputThenParams()](https://github.com/eclipse/jetty.project/commit/c70fa88be038dc6868a8bb271d184d87ac8112b2#diff-37eb9749f0598ab583e805f8fcedd5eeR837) that can replicate this scenario.

@janbartel what should the proper fix be for this?

**Comment by janbartel** [comment]
@joakime fundamentally either the application reads the input stream, or the container reads the input stream because of a MultipartConfig annotation or equivalent in web.xml.  You can't have both.

**Comment by janbartel** [comment]
@joakime  there are 2 tests for the multipart stuff that tests whether the input stream has already been consumed or not:

https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-servlets/src/test/java/org/eclipse/jetty/servlets/MultipartFilterTest.java#L748

and

https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-util/src/test/java/org/eclipse/jetty/util/MultiPartInputStreamTest.java#L244

Both of those tests are passing.  Please modify the tests to reproduce an exception being thrown instead of returning 0 parts.

**Comment by joakime** [comment]
@janbartel here's the testcase (currently set to @Ignore) ..

https://github.com/eclipse/jetty.project/blob/jetty-9.3.x/jetty-server/src/test/java/org/eclipse/jetty/server/RequestTest.java#L836-L837

**Comment by janbartel** [comment]
@joakime If you're just doing a POST and the input has already been consumed, then you won't get an exception. You won't get any content params either.

 The MultipartException only occurs iff you have declared a multipart config in an annotation or web.xml && you have a filter that consumes the input before the container && you receive a multipart/form-data request.  I think this situation is  a wrong configuration of the webapp, so I think the specific multipart exception shows exactly where the problem lies.  If we were to make it the same behaviour as POST params, then there'd be no error no params - which won't help you diagnose what's gone wrong.  OTOH if we throw an IllegalStateException instead, I'm not sure that's any more informative than the MultipartException?

**Comment by janbartel** [comment]
I'm going to close this issue:   as I've previously stated, I think the problem here is that you cannot try and consume multipart input twice - you can't have both a filter that reads all the content AND have an annotation that says you're expecting the container to consume it as well.

If anyone has a strong disagreement, and a suggestion for a different way of handling this misconfiguration, then please feel free to reopen this issue and explain your proposition in detail.

**Comment by joakime** [comment]
This is a complex issue, and on contemplation, its really 2 different backwards compat issues.

If you have an existing (legacy) webapp developed against Servlet 2.5, using multipart/form-data form submission will result in valid parameters without the need for a MultiPartConfig setup.

When the move to Servlet 3.0 occurred, the multipart/form-data now was gated behind the existence of MultiPartConfig on the servlet too. (I think this is wrong, and is breaking backward compat with older webapps).

If the webapp also happens to use commons-fileupload, which works on older Servlet 2.5 webapps and containers, this setup will now fail on Jetty 9.

So in short...

 * Servlet 2.5 behavior with request.getParameters() and multipart/form-data is missing in Jetty 9.  (this is because Jetty 9 gates multipart/form-data behind the need for a declared MultipartConfig)
 * Using Servlet 3.0 behaviors in combination with commons-fileupload is unsupported in Jetty 9 (a shocking behavior).

Fixing this is 2 fold.

1) the request.getParameters() on multipart/form-data should support getting parameters without MultipartConfig existence (the Servlet 2.x behavior)
2) The request.getParameters() extraction of empty/EOF request input streams should not throw an exception.

**Comment by joakime** [comment]
I'll try to prepare an example servlet 2.5 war to flesh out the backwards compat behavior differences.
Reopening and assigning to myself for this sample war step.

**Comment by janbartel** [comment]
@joakime:

1.  the old servlet 2.5 behaviour is absolutely supported: if your webapp used the old MultipartFilter, you can absolutely continue to use it.
2. the new servlet 3.0 @MultiPart annotations are absolutely supported. If you upgrade your webapp to use servlet 3.0, we fully support you.
3. the only limitation is that you cannot do both 1. and 2. together at the same time - because it makes no sense! You can't install a filter to consume the input and also define a servlet that expects to consume that same input.

**Comment by joakime** [comment]
Closing as not valid, per clarification by Jan.

For those that arrived here via a google search:

Prior to Servlet 3.0, a `multipart/form-data` submitted content was not handled by the Servlet spec.
You would need to parse the Request body content using a non-servlet library.
For Jetty, we offered a `MultipartFilter` to do the parsing work.
But there was also the `commons-fileupload` library, which can also do the parsing work.

If you used `multipart/form-data` prior to Servlet 3.0 (even if you didn't specify files to be uploaded in your form) one of those 2 libraries would be used to access the form parameters.

In environments prior to Servlet 3.0, using `HttpServletRequest.getParameter*()` methods to access the form parameters from a request body content that used `multipart/form-data` would result in no parameters.

With Servlet 3.0 you had `multipart/form-data` support built into the servlet-spec with the addition of `MultipartConfig` and `HttpServletRequest.getParts()`

In Servlet 3.0+ environments using `HttpServletRequest.getParameter*()` methods to access the form parameters from a request body content that used `multipart/form-data` would result in valid parameters, but only if the servlet had a `@MultipartConfig` setup for it.

If you had a old webapp, using Servlet 2.5 (for example), that still used `commons-fileupload` internally, but you deployed it on Jetty 9 (a Servlet 3.1 environment), you would still see the Servlet 2.5 behavior, no `multipart/form-data` auto parsing, and the need to use `commons-fileupload` (or similar) to access the form parameters (including the optional files being uploaded).
